### DISCUSSION TITLE
Spatial Reference for unstructured grid in class PstFrom?

### DISCUSSION THREAD

**Discussion by g3rm7** [gmdsi/GMDSI_notebooks#102] [Q&A] [answered]
Hi, I trying to create a pst from scratch for an unstructured grid model, following the steps proposed in the "freyberg_pstfrom_pest_setup" notebook, and I don't know what to use as "spatial_reference" variable in the "PstFrom" class instantiating. I understand that this is required for later parameter covariances definitions. 

In the notebook, the class "pyemu.helpers.SpatialReference" is used for creating the "spatial_reference", but I understand that this only works for structured grid. It is mentioned that the flopy "modelgrid" instance could be used for this, but I'm not sure how to use it to define the "spatial_reference" variable, or what's the structure that this variable should have

Can anyone hint me on how this should be done? 
Regards!

**Comment by laat0003** [comment]
Howdy @g3rm7,

I'm not a dev here but I am currently working on a MF6 model using DISV. I didn't need to create a spatial reference object but just passed in the modelgrid object from Flopy as the spatial reference when instantiating PstFrom and that worked like a charm.

**Reply by rhugman** [reply, marked as answer]
> Howdy @g3rm7,
> 
> 
> 
> I'm not a dev here but I am currently working on a MF6 model using DISV. I didn't need to create a spatial reference object but just passed in the modelgrid object from Flopy as the spatial reference when instantiating PstFrom and that worked like a charm.

This will work for DISV, but not DISU. I think there is a workaround (mentioned by @jtwhite79 during the session?)...but I avoid complicated things like DISU whenever possible, so I dunno :) 

All of this becomes moot once pypesutils is ready.

**Reply by g3rm7** [reply]
Thanks for your answers!
As @rhugman commented, in my case I was working with a modflow-USG model, using DISU. Also tried passing directly the "modelgrid", but didn't worked...
 
- Does anyone know what's the workaround mentioned by @rhugman or @jtwhite79 ?
- I had some problems too, when trying to set the K parametrization based on the external files. How this need to be done in USG, considering that I have a single K file per layer? 
- What is the pypesutils and how that could help with this issue? 
Thanks

**Reply by rhugman** [reply]
I think this is the workaround, but Ive never tried to implement it myself: https://github.com/pypest/pyemu/issues/306

Pypestutils (in alpha, so user beware): https://github.com/pypest/pypestutils

No idea about MFUSG, sorry!

**Reply by laat0003** [reply]
I've implemented it a couple of times but I do something different with zones to what is in the test example. I used zones to define my layers. I passed in separate dataframes for pilot-points (ppspace option) in each layer. The dataframes for each layer were made using geopandas by intersecting a regular grid of points (based on model extent) with a polygon of the specific layer (dissolve the geometry of the layer via geopandas dataframe).

The first model I did this with had 35 "layers" and pinch-outs all over the place plus zones in layers (can be problematic if you are already using zones to define layers) so was a real pain.  Where I work currently almost all models are somewhat overly complex (in my opinion) MF-USG models. If I get the chance to build anew then I jump to DISV in MF6 and use vertical pass through via IDOMAIN for pinch-outs if I have to.

**Reply by mnfienen** [reply]
hey @laat0003 - I agree that some pretty complicated geometric pinch-out layering issues can be sorted using DISV and IDOMAIN=-1 pass through layers with MF6. Can save significant hassle
